### PR TITLE
Bindu | RESTWS-888 | Add support for program Attribute Type Concept

### PR DESCRIPTION
**Pull request by binduak**
[RESTWS-888](https://issues.openmrs.org/browse/RESTWS-888) Add support for program attribute of type "concept" for ProgramAttributeTypeResource

### CODE REVIEW

**Comment by dkayiwa** [comment]
@binduak do you mind adding some tests for this?

**Comment by binduak** [comment]
> @binduak do you mind adding some tests for this?

sure @dkayiwa

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java [hunk 1]
```diff
@@ -39,31 +46,47 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
...
 		return new NeedsPaging<ProgramAttributeType>(Context.getProgramWorkflowService()
 		        .getAllProgramAttributeTypes(),
 		        context);
 	}
+
+	@PropertyGetter("concept")
+	public Object getConcept(ProgramAttributeType delegate) {
+		if (!OpenmrsUtil.nullSafeEquals(delegate.getDatatypeClassname(), ConceptDatatype.class.getCanonicalName()))
+			return null;
+		else {
```
Comment (gsluthra) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java:81]: The else is not needed. The return statement on top exists this method if the condition is true.
You can now write the remaining logic without the else { } .

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java [hunk 2]
```diff
@@ -39,31 +46,47 @@ public DelegatingResourceDescription getRepresentationDescription(Representation
...
+			Concept concept;
+			String id = delegate.getDatatypeConfig();
+			if (Pattern.compile("\\d+").matcher(id).matches()) {
```
Comment (gsluthra) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java:84]: Is this code written to check if id is a valid Integer? If so, then there might already be a utils method for this, which is more performant and clearer than regex. Also some related reading: https://www.baeldung.com/java-check-string-number (provides alternatives on Apache Commons library for this purpose). Might make for clearer code.

Comment (binduak) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java:83]: @gsluthra Thought of having this in a util file but i don't see any generic util file in the webservices.rest module. The other option could be separating the if condition into a private method in the same class. Please suggest what can be the best option

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java (hunk 1 above)
Comment (binduak) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java]: @gsluthra addressed the comment

**Comment by binduak** [comment]
> > @binduak do you mind adding some tests for this?
> 
> sure @dkayiwa

@dkayiwa added test cases. please check. Thanks !

**Comment by dkayiwa** [comment]
@binduak do you mind updating the JIRA ticket description with an example of a REST call that you would like to have with this functionality?

**Comment by binduak** [comment]
> @binduak do you mind updating the JIRA ticket description with an example of a REST call that you would like to have with this functionality?

done @dkayiwa

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java (hunk 2 above)
Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java:83]: How about the Apache Commons library?

Comment (dkayiwa) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java:83]: @binduak what do you think of the above comment?

**Comment by jwnasambu** [comment]
LGTM!!

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java (hunk 2 above)
Comment (binduak) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/resource/openmrs2_2/ProgramAttributeTypeResource2_2.java]: @dkayiwa handled the feedback changes as per the suggestions

**Comment by binduak** [comment]
Thanks @dkayiwa

**Pull request merged**